superbench/benchmarks/micro_benchmarks/gpu_sm_copy_performance/gpu_sm_copy.cu

```diff
@@ -0,0 +1,297 @@
+// Copyright (c) Microsoft Corporation.
+// Licensed under the MIT License.
+
+// GPU SM copy benchmark tests dtoh/htod data transfer bandwidth initiated by GPU SM.
+
+#include <chrono>
+#include <cstdio>
+#include <cstring>
+#include <string>
+
+#include <cuda.h>
+#include <cuda_runtime.h>
+
+// Argurment index used in argument parsing.
+enum class ArgIdx { kGpuId = 1, kCopyDirection, kSize, kNumLoops, kNumArgs };
+
+// ID of GPU used in this benchmark.
+int g_gpu_id = 0;
+
+// Data transfer direction, can be "dtoh" or "htod".
+std::string g_copy_direction;
+
+// Data buffer size used.
+uint64_t g_size = 0;
+
+// Number of loops in data transfer benchmark.
+uint64_t g_num_loops = 0;
+
+// Original data buffer.
+uint8_t *g_data_buf = nullptr;
+
+// Buffer to validate the correctness of data transfer.
+uint8_t *g_check_buf = nullptr;
+
+// Data buffer in host memory.
+uint8_t *g_host_buf = nullptr;
+
+// Device pointer of the data buffer in host memory.
+uint8_t *g_host_buf_dev_ptr = nullptr;
+
+// Data buffer in device memory
+uint8_t *g_dev_buf = nullptr;
+
+// Pring usage of this program.
+void PrintUsage() { printf("Usage: gpu_sm_copy <gpu-id> <copy-direction: dtoh|htod> <size> <num_loops>\n"); }
+
+// Prepare data buffers to be used.
+int PrepareBuf() {
+    cudaError_t cuda_err = cudaSuccess;
+
+    // Generate data to copy
+    g_data_buf = static_cast<uint8_t *>(malloc(g_size));
+    for (int i = 0; i < g_size; i++) {
+        g_data_buf[i] = static_cast<uint8_t>(i % 256);
+    }
+
+    // Reset check buffer
+    g_check_buf = static_cast<uint8_t *>(malloc(g_size));
+    for (int i = 0; i < g_size; i++) {
+        g_check_buf[i] = static_cast<uint8_t>(0);
+    }
+
+    // Allocate host buffer
+    g_host_buf = static_cast<uint8_t *>(malloc(g_size));
+    cuda_err = cudaHostRegister(g_host_buf, g_size, cudaHostRegisterMapped);
+    if (cuda_err != cudaSuccess) {
+        fprintf(stderr, "PrepareBuf::cudaHostRegister error: %d\n", cuda_err);
+        return -1;
+    }
+    cuda_err = cudaHostGetDevicePointer((void **)&g_host_buf_dev_ptr, g_host_buf, 0);
+    if (cuda_err != cudaSuccess) {
+        fprintf(stderr, "PrepareBuf::cudaHostGetDevicePointer error: %d\n", cuda_err);
+        return -1;
+    }
+
+    // ALlocate device buffer
+    cuda_err = cudaMalloc(&g_dev_buf, g_size);
+    if (cuda_err != cudaSuccess) {
+        fprintf(stderr, "PrepareBuf::cudaMalloc error: %d\n", cuda_err);
+        return -1;
+    }
+
+    // Initialize source buffer
+    if (g_copy_direction == "dtoh") {
+        cuda_err = cudaMemcpy(g_dev_buf, g_data_buf, g_size, cudaMemcpyDefault);
```

Comment:
for dtoh, I thought the destination should be host, i.e. the `cudaMemcpy(g_host_buf,...)`